src/vellum/workflows/nodes/experimental/tool_calling_node/node.py

```diff
@@ -0,0 +1,127 @@
+from vellum.workflows.graph.graph import Graph
+from vellum.workflows.inputs.base import BaseInputs
+from vellum.workflows.nodes.bases import BaseNode
+from vellum.workflows.nodes.experimental.tool_calling_node.utils import (
+    ToolRouterNode,
+    create_function_node,
+    create_tool_router_node,
+)
+from vellum.workflows.outputs.base import BaseOutputs
+from vellum.workflows.state.base import BaseState
+from vellum.workflows.state.context import WorkflowContext
+from vellum.workflows.types.core import EntityInputsInterface, MergeBehavior
+from vellum.workflows.workflows.base import BaseWorkflow
+
+
+class ToolCallingNode(BaseNode):
+    """
+    A Node that allows for tool calling while dynamically generating a graph based on provided functions.
+
+    Attributes:
+        ml_model: str - The model to use for tool calling (e.g., "gpt-4o-mini")
+        blocks: List[PromptBlock] - The prompt blocks to use (same format as InlinePromptNode)
+        functions: List[FunctionDefinition] - The functions that can be called
+        function_callables: List[Callable] - The callables that can be called
+        prompt_inputs: Optional[EntityInputsInterface] - Mapping of input variable names to values
+    """
+
+    ml_model: ClassVar[str] = "gpt-4o-mini"
+    blocks: ClassVar[List[PromptBlock]] = []
+    functions: ClassVar[List[FunctionDefinition]] = []
+    function_callables: ClassVar[List[Callable]] = []
+    prompt_inputs: ClassVar[Optional[EntityInputsInterface]] = None
+
+    class Trigger(BaseNode.Trigger):
+        merge_behavior = MergeBehavior.AWAIT_ANY
+
+    class Outputs(BaseOutputs):
+        """
+        The outputs of the ToolCallingNode.
+
+        text: The final text response after tool calling
+        chat_history: The complete chat history including tool calls
+        """
+
+        text: str = ""
+        chat_history: List[ChatMessage] = []
+
+    def run(self) -> Outputs:
+        """
+        Run the tool calling workflow.
+
+        This dynamically builds a graph with router and function nodes,
+        then executes the workflow.
+        """
+        # TODO: We should parse input values and add them to the state
+        self._build_graph()
+
+        with execution_context(parent_context=get_parent_context()):
+
+            class ToolCallingState(BaseState):
+                chat_history: List[ChatMessage] = []
+
+            class ToolCallingWorkflow(BaseWorkflow[BaseInputs, ToolCallingState]):
+                graph = self._graph
+
+                class Outputs(BaseWorkflow.Outputs):
+                    text: str = ToolRouterNode.Outputs.text
+                    chat_history: List[ChatMessage] = ToolCallingState.chat_history
+
+            subworkflow = ToolCallingWorkflow(
+                parent_state=self.state,
+                context=WorkflowContext(
+                    vellum_client=self._context.vellum_client, generated_files=self._context.generated_files
+                ),
```

Comment:
[nit] `WorkflowContext` should now have a static method for initiating new workflow context's